In [1]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 2: Corruption Checksum ---</h2><p>As you walk through the door, a glowing humanoid shape yells in your direction. "You there! Your state appears to be idle. Come help us repair the corruption in this spreadsheet - if we take another millisecond, we'll have to display an hourglass cursor!"</p>
<p>The spreadsheet consists of rows of apparently-random numbers. To make sure the recovery process is on the right track, they need you to calculate the spreadsheet's <em>checksum</em>. For each row, determine the difference between the largest value and the smallest value; the checksum is the sum of all of these differences.</p>
<p>For example, given the following spreadsheet:</p>
<pre><code>5 1 9 5
7 5 3
2 4 6 8</code></pre>
<ul>
<li>The first row's largest and smallest values are <code>9</code> and <code>1</code>, and their difference is <code>8</code>.</li>
<li>The second row's largest and smallest values are <code>7</code> and <code>3</code>, and their difference is <code>4</code>.</li>
<li>The third row's difference is <code>6</code>.</li>
</ul>
<p>In this example, the spreadsheet's checksum would be <code>8 + 4 + 6 = 18</code>.</p>
<p><em>What is the checksum</em> for the spreadsheet in your puzzle input?</p>
</article>


In [21]:
from functools import reduce
from operator import sub
from re import findall

from more_itertools import minmax
from tabulate import tabulate


example = """
5 1 9 5
7 5 3
2 4 6 8
"""


def checksum(spreadsheat: str) -> int:
    return sum(
        reduce(lambda v, e: e - v, minmax(map(int, findall(r"\d+", line))), 0)
        for line in spreadsheat.strip().splitlines()
    )


assert checksum(example) == 18

In [22]:
with open("../input/day2.txt") as f:
    puzzle = f.read()

print(f"Part I: {checksum(puzzle)}")

Part I: 41919


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>41919</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>"Great work; looks like we're on the right track after all.  Here's a <em class="star">star</em> for your effort." However, the program seems a little worried. Can programs <em>be</em> worried?</p>
<p>"Based on what we're seeing, it looks like all the User wanted is some information about the <em>evenly divisible values</em> in the spreadsheet.  Unfortunately, none of us are equipped for that kind of calculation - most of us specialize in <span title="Bonus points if you solve this part using only bitwise operations.">bitwise operations</span>."</p>
<p>It sounds like the goal is to find the only two numbers in each row where one evenly divides the other - that is, where the result of the division operation is a whole number. They would like you to find those numbers on each line, divide them, and add up each line's result.</p>
<p>For example, given the following spreadsheet:</p>
<pre><code>5 9 2 8
9 4 7 3
3 8 6 5</code></pre>
<ul>
<li>In the first row, the only two numbers that evenly divide are <code>8</code> and <code>2</code>; the result of this division is <code>4</code>.</li>
<li>In the second row, the two numbers are <code>9</code> and <code>3</code>; the result is <code>3</code>.</li>
<li>In the third row, the result is <code>2</code>.</li>
</ul>
<p>In this example, the sum of the results would be <code>4 + 3 + 2 = 9</code>.</p>
<p>What is the <em>sum of each row's result</em> in your puzzle input?</p>
</article>

</main>


In [37]:
from more_itertools import one


example_II = """
5 9 2 8
9 4 7 3
3 8 6 5
"""


def evenly_divisible_checksum(spreadsheat: str) -> int:
    return sum(
        one(
            a // b
            for a, b in product(map(int, findall(r"\d+", line)), repeat=2)
            if a != b and a % b == 0
        )
        for line in spreadsheat.strip().splitlines()
    )


assert evenly_divisible_checksum(example_II) == 9

In [38]:
print(f"Part II: {evenly_divisible_checksum(puzzle)}")

Part II: 303


<link href="style.css" rel="stylesheet"></link>

<p>Your puzzle answer was <code>303</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
